### PR TITLE
feat: CIS spread sheet to OSCAL catalog

### PR DESCRIPTION
**Pull request by degenaro**
Signed-off-by: degenaro <lou.degenaro@gmail.com>

## Types of changes

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [ ] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

**Comment by degenaro** [comment]
@enikonovad This one is ready for review. @vikas-agarwal76 has looked over the produced catalogs and has given his seal of approval.

File: trestle/tasks/cis_xlsx_to_oscal_catalog.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+        if not append:
+            control.props.append(prop)
+
+    def _chk_prop(self, control: Control, prop: Property) -> None:
+        """Check property in control."""
+        for control_prop in control.props:
+            if prop.name == control_prop.name:
+                if prop.value != control_prop.value:
+                    text = f'{control.id} property {prop.name} values {prop.value} and {control_prop.value} not equal'
+                    raise RuntimeError(text)
```
Comment (enikonovad) [inline trestle/tasks/cis_xlsx_to_oscal_catalog.py:162]: Use TrestleError

File: trestle/tasks/cis_xlsx_to_oscal_catalog.py [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+            # group or control
+            if recommendation is None:
+                catalog_helper.add_group(section, title, props, parts)
+            else:
+                self._add_property_boolean(xlsx_helper, props, row, 'v7 IG1')
+                self._add_property_boolean(xlsx_helper, props, row, 'v7 IG2')
+                self._add_property_boolean(xlsx_helper, props, row, 'v7 IG3')
+                self._add_property_boolean(xlsx_helper, props, row, 'v8 IG1')
+                self._add_property_boolean(xlsx_helper, props, row, 'v8 IG2')
+                self._add_property_boolean(xlsx_helper, props, row, 'v8 IG3')
```
Comment (enikonovad) [inline trestle/tasks/cis_xlsx_to_oscal_catalog.py:379]: Please remove duplicates

**Review by enikonovad — CHANGES_REQUESTED** [review]
Overall looks good, please decrease number of duplicates

File: trestle/tasks/cis_xlsx_to_oscal_catalog.py [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+                group.parts = parts
+            parent.groups.append(group)
+            self._all_groups[section] = group
+
+    def _add_prop(self, control: Control, prop: Property) -> None:
+        """Add property to control."""
+        control_props = control.props
+        control.props = []
+        match = 0
+        append = 0
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/cis_xlsx_to_oscal_catalog.py]: Fixed.

File: trestle/tasks/cis_xlsx_to_oscal_catalog.py (hunk 1 above)
Comment (degenaro) [reply trestle/tasks/cis_xlsx_to_oscal_catalog.py]: Fixed.

(Removed as extraneous.)

File: trestle/tasks/cis_xlsx_to_oscal_catalog.py (hunk 2 above)
Comment (degenaro) [reply trestle/tasks/cis_xlsx_to_oscal_catalog.py:393]: These are not duplicates. Each function all has a unique set of parameters.

File: tests/trestle/tasks/cis_xlsx_to_oscal_catalog_test.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+
+def test_cis_xlsx_to_oscal_catalog_execute_ocp(tmp_path: pathlib.Path):
+    """Test execute call - ocp."""
+    config = configparser.ConfigParser()
+    config_path = pathlib.Path('tests/data/tasks/cis-xlsx-to-oscal-catalog/test-cis-xlsx-to-oscal-catalog.ocp.config')
+    config.read(config_path)
+    section = config['task.cis-xlsx-to-oscal-catalog']
+    section['output-dir'] = str(tmp_path)
+    tgt = cis_xlsx_to_oscal_catalog.CisXlsxToOscalCatalog(section)
```
[the remark this answers is not in the document]
Comment (degenaro) [reply tests/trestle/tasks/cis_xlsx_to_oscal_catalog_test.py]: Fixed.

File: trestle/tasks/cis_xlsx_to_oscal_catalog.py [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+        self._root_group = OrderedDict()
+        self._all_groups = OrderedDict()
+        self._all_controls = OrderedDict()
+
+    def add_group(self, section: str, title: str, props: List[Property], parts: List[Part]) -> None:
+        """Add group."""
+        numdots = section.count('.')
+        if numdots == 0:
+            group = Group(title=f'{title}', id=f'CIS-{section}')
+            if len(props):
```
[the remark this answers is not in the document]
Comment (degenaro) [reply trestle/tasks/cis_xlsx_to_oscal_catalog.py]: Fixed.

**Comment by degenaro** [comment]
@enikonovad Thx for the review. My claim is that I have addressed all requested changed.

File: trestle/tasks/cis_xlsx_to_oscal_catalog.py [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+    def _add_prop(self, control: Control, prop: Property) -> None:
+        """Add property to control."""
+        control_props = control.props
+        control.props = []
+        first = -1
+        last = -1
+        for i, control_prop in enumerate(control_props):
+            if control_prop.name == prop.name:
+                if first < 0:
+                    first = i
```
Comment (enikonovad) [inline trestle/tasks/cis_xlsx_to_oscal_catalog.py:149]: `first` is not needed?

**Review by enikonovad — APPROVED** [review]
LGTM - a minor comment

File: trestle/tasks/cis_xlsx_to_oscal_catalog.py (hunk 6 above)
Comment (degenaro) [reply trestle/tasks/cis_xlsx_to_oscal_catalog.py]: You are correct. Fixed.

**Pull request merged**